praw/exceptions.py

```diff
@@ -7,32 +7,85 @@
 All other exceptions are subclassed from :class:`.ClientException`.
 
 """
-from typing import Optional
+from typing import List, Optional, TypeVar, Union
+
+_RedditErrorItem = TypeVar("_RedditErrorItem")
 
 
 class PRAWException(Exception):
     """The base PRAW Exception that all other exception classes extend."""
 
 
-class APIException(PRAWException):
-    """Indicate exception that involve responses from Reddit's API."""
+class RedditErrorItem:
+    """Represents a single error returned from Reddit's API."""
 
-    def __init__(self, error_type: str, message: str, field: Optional[str]):
-        """Initialize an instance of APIException.
+    @property
+    def error_message(self):
+        """Get the completed error message string."""
+        error_str = "{}: '{}'".format(self.error_type, self.message)
+        if self.field:
+            error_str += " on field '{}'".format(self.field)
+        return error_str
+
+    def __init__(
+        self, error_type: str, message: str, field: Optional[str] = None
```

Comment:
In that case why don't I deal with it in the `RedditErrorItem`?

Reply:
The change you made is OK I guess, because it’s not much in this case.

But I don’t feel it’s optimal, I think it’s better to fully abstract away the logic used to refine the API data. Constructors should be simple and just take exactly whatever they need for the object to work. Especially a data class, they’re meant to just hold exactly whatever they’re given, that’s it’s only responsibility.

I personally would write a function to construct the object from the list the API provides. If the structure were more complicated or susceptible to change you’d only need to change the logic in the function and not any part of the class.